Type updates

Sheyne opened the pull request:
Types is now rewritten from the ground up.

Classes now support (multiple) inheritance. 
   - multiple inheritance is a good thing because what I am calling classes are really just interfaces

Types can now be parsed from literals

```typescript
Type.parseType("class { x: String }")
Type.parseType("List<Number>")
Type.parseScope(`
DFANode = class Node {
    AcceptState: Boolean
}
`)
```

Sheyne commented on package.json:
```diff
@@ -15,7 +15,7 @@
      "peg": "pegjs --allowed-start-rules Definitions,Type --dependency \"./type-structures\" app/types/type-parser.pegjs && mv app/types/type-parser.js app/types/type-parser-inner.js && sed -i \"\" 's/^var \\.\\/type-structures/var types/' app/types/type-parser-inner.js",
```
I realize that this SED script is ridiculous, not sure there's a better way for now. 

This currently copies the '.js' file directly into the "app/" folder. @RighteousRaichu do you know how to fix this?

RighteousRaichu replied:
`"pegjs --allowed-start-rules Definitions,Type --dependency \"types:./type-structures\" app/types/type-parser.pegjs && mv app/types/type-parser.js app/types/type-parser-inner.js"`

Is this what you're asking for?

And what do you mean it copies it to the "app/" folder? It looks like the mv command just renames it?

Sheyne commented on app/components/properties-panel/properties-panel.component.html:
```diff
@@ -33,23 +26,32 @@
+                                </label>
+                            </template>
+
+                            <!-- add neither case-->
```
Maybe import the elif package?

slaymaker1907 commented on app/types/type-parser.pegjs:
```diff
@@ -0,0 +1,66 @@
+// This represents a scope
+// currently it doesn't make any sense to have a nested scope
+
+Definitions = _ "{" _ dfns:Definition* _ "}" _ { return new types.TypeScope(new Map(dfns)); }
```
If we are making our own language,

```
return new types.TypeScope(new Map(dfns));
```
Seems much too verbose. Maintaining similarity to JavaScript is good, but we should take advantage of having our own parser where possible. For instance, at the very least we should remove the both "new"s from the spec.

slaymaker1907 replied:
With this comment and any others I make, please let me know if I'm just misunderstanding this parser generator.

Sheyne replied:
If you read the app/types/README.md that I just committed, there are several examples of syntax. 

Link to file in the commit [here](https://github.com/2graphic/sinap-ide/blob/5f8d11e4786862c1b781126de619dca168137984/app/types/README.md). (This directly references the commit so people from the future be warned: it might of changed)

slaymaker1907 commented on app/types/type-structures.ts:
```diff
@@ -0,0 +1,317 @@
+export class TypeVariable implements Type {
+    kind = "Variable";
+
+    constructor(public type?: Type, public lookupName?: string, public matchName?: string) {
```
Why are all of these optional? It seems like these should be required fields intuitively. If partial initialization is needed, JSON may be more appropriate until fully initialized.

Sheyne replied:
TypeVariables are used in two contexts, type inference, and type referencing. Either can define a match name. All three optional can be defined, or none of them, things will still work.

slaymaker1907 commented on app/types/type-structures.ts:
```diff
@@ -0,0 +1,317 @@
+        return false;
+    }
+
+    rsubtype(that: Type) {
```
Personally, I would have subtypeing return a new type so that you still have access to the old one if needed and it seems like it wouldn't be too much of a performance hit.

I also don't like combining read and write methods.

Sheyne replied:
Subtype is not a create method, it (as wasn't documented when you commented but is now) tests whether `this` is a subtype of `that`. Mutation only happens if it is a type variable in type inference. `rsubtype` is the reverse, in principle: 

```TypeScript
this.subtype(that) === that.subtype(this)
```
Essentially I want things that define a `rsubtype` to have more say in the process.

slaymaker1907 commented on app/types/type-structures.ts:
```diff
@@ -0,0 +1,317 @@
+    validate() { }
+
+    toString() {
+        return "(" + this.matchName + "?" + this.lookupName + ":" + this.type + ")";
```
Maybe change this out with an ES6 template string for clarity.

slaymaker1907 commented on app/types/type-structures.ts:
```diff
@@ -0,0 +1,317 @@
+
+    subtype(that: Type): boolean {
+        if (that.rsubtype) {
+            return that.rsubtype(this);
```
Couldn't this potentially mutate that?

Sheyne replied:
Yes, but only if its a type variable.

slaymaker1907 commented on app/types/type-structures.ts:
```diff
@@ -0,0 +1,317 @@
+    validate() { }
+}
+
+export class ThunkType implements Type {
```
What is a Thunk?

Sheyne replied:
Zero argument function, used to delay execution. Will probably be used for computed fields.

slaymaker1907 commented on package.json:
```diff
@@ -12,9 +12,10 @@
     "clean": "rm -r ./build 2> /dev/null || echo > /dev/null && rm -r ./dist 2> /dev/null || echo > /dev/null",
     "clean:node_modules": "npm run clean && rm -r ./node_modules 2> /dev/null || echo > /dev/null",
     "webpack": "webpack --progress --profile --colors --display-error-details --display-cached",
+     "peg": "pegjs --allowed-start-rules Definitions,Type --dependency \"./type-structures\" app/types/type-parser.pegjs && mv app/types/type-parser.js app/types/type-parser-inner.js && sed -i \"\" 's/^var \\.\\/type-structures/var types/' app/types/type-parser-inner.js",
```
This should have a windows version if possible.

Sheyne commented:
@RighteousRaichu will you read through the inline comments I added and the README.md and let me know whether you think that's sufficient? 

Also @slaymaker1907 I think that I've addressed most of your confusion about syntax in the readme.
